diff --git a/utils/bazel/llvm-project-overlay/libc/BUILD.bazel b/utils/bazel/llvm-project-overlay/libc/BUILD.bazel
--- a/utils/bazel/llvm-project-overlay/libc/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/libc/BUILD.bazel
@@ -32,7 +32,7 @@
 # Flag documentation: https://bazel.build/extending/config
 string_flag(
     name = "mpfr",
-    build_setting_default = "external",
+    build_setting_default = "disable",
     values = [
         "disable",  # Skip tests that need mpfr
         "external",  # Build mpfr from source
diff --git a/utils/bazel/llvm-project-overlay/lld/BUILD.bazel b/utils/bazel/llvm-project-overlay/lld/BUILD.bazel
--- a/utils/bazel/llvm-project-overlay/lld/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/lld/BUILD.bazel
@@ -106,8 +106,8 @@
         "//llvm:TargetParser",
         "//llvm:TransformUtils",
         "//llvm:config",
-        "//third-party:zlib",
-        "//third-party:zstd",
+        # "//third-party:zlib",
+        # "//third-party:zstd",
     ],
 )
 
diff --git a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
--- a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
@@ -324,14 +324,8 @@
         ":config",
         ":Demangle",
         "//third-party/siphash",
-        # We unconditionally depend on the custom LLVM zlib wrapper. This will
-        # be an empty library unless zlib is enabled, in which case it will
-        # both provide the necessary dependencies and configuration defines.
-        "//third-party:zlib",
-        # We unconditionally depend on the custom LLVM zstd wrapper. This will
-        # be an empty library unless zstd is enabled, in which case it will
-        # both provide the necessary dependencies and configuration defines.
-        "//third-party:zstd",
+        # "//third-party:zlib",
+        # "//third-party:zstd",
     ],
 )
 
@@ -4434,7 +4428,7 @@
 # Flag documentation: https://bazel.build/extending/config
 string_flag(
     name = "pfm",
-    build_setting_default = "external",
+    build_setting_default = "disable",
     values = [
         "disable",  # Don't include pfm at all
         "external",  # Build pfm from source
