diff --git a/clang/include/clang/Basic/Attr.td b/clang/include/clang/Basic/Attr.td
--- a/clang/include/clang/Basic/Attr.td
+++ b/clang/include/clang/Basic/Attr.td
@@ -5345,3 +5345,33 @@
   let Subjects = SubjectList<[Function]>;
   let Documentation = [ModularFormatDocs];
 }
+
+// Begin TAPA
+
+def TapaPipeline : Attr {
+  let Spellings = [GNU<"tapa_pipeline">,
+                   CXX11<"tapa","pipeline">,
+                   C23<"tapa", "pipeline">];
+  let Args = [IntArgument<"II", /* optional */1>];
+  let Documentation = [Undocumented];
+}
+
+def TapaUnroll : Attr {
+  let Spellings = [GNU<"tapa_unroll">,
+                   CXX11<"tapa","unroll">,
+                   C23<"tapa", "unroll">];
+  let Args = [IntArgument<"Factor", /* optional */1>];
+  let Documentation = [Undocumented];
+}
+
+def TapaTarget : Attr {
+  let Spellings = [GNU<"tapa_target">,
+                   CXX11<"tapa","target">,
+                   C23<"tapa", "target">];
+  let Args = [EnumArgument<"Target", "TargetType", /*is_string=*/true,
+                           ["xilinx_hls", "xilinx_vitis", "ignore"],
+                           ["XilinxHLS", "XilinxVitis", "Ignore"]>];
+  let Documentation = [Undocumented];
+}
+
+// End TAPA
diff --git a/clang/include/clang/Basic/AttributeCommonInfo.h b/clang/include/clang/Basic/AttributeCommonInfo.h
--- a/clang/include/clang/Basic/AttributeCommonInfo.h
+++ b/clang/include/clang/Basic/AttributeCommonInfo.h
@@ -72,7 +72,7 @@
     IgnoredAttribute,
     UnknownAttribute,
   };
-  enum class Scope { NONE, CLANG, GNU, MSVC, OMP, HLSL, VK, GSL, RISCV };
+  enum class Scope { NONE, CLANG, GNU, MSVC, OMP, HLSL, VK, GSL, RISCV, TAPA };
   enum class AttrArgsInfo {
     None,
     Optional,
diff --git a/clang/lib/Basic/Attributes.cpp b/clang/lib/Basic/Attributes.cpp
--- a/clang/lib/Basic/Attributes.cpp
+++ b/clang/lib/Basic/Attributes.cpp
@@ -236,7 +236,8 @@
       .Case("vk", AttributeCommonInfo::Scope::VK)
       .Case("msvc", AttributeCommonInfo::Scope::MSVC)
       .Case("omp", AttributeCommonInfo::Scope::OMP)
-      .Case("riscv", AttributeCommonInfo::Scope::RISCV);
+      .Case("riscv", AttributeCommonInfo::Scope::RISCV)
+      .Case("tapa", AttributeCommonInfo::Scope::TAPA);
 }
 
 unsigned AttributeCommonInfo::calculateAttributeSpellingListIndex() const {
diff --git a/clang/lib/Sema/SemaDeclAttr.cpp b/clang/lib/Sema/SemaDeclAttr.cpp
--- a/clang/lib/Sema/SemaDeclAttr.cpp
+++ b/clang/lib/Sema/SemaDeclAttr.cpp
@@ -162,6 +162,43 @@
   return checkStringLiteralArgumentAttr(AL, ArgExpr, Str, ArgLocation);
 }
 
+// Begin TAPA
+
+static void handleTapaPipelineAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  uint32_t ii = 0;
+  if (AL.getNumArgs()) {
+    if (!S.checkUInt32Argument(AL, AL.getArgAsExpr(0), ii)) return;
+  }
+
+  D->addAttr(::new (S.Context) TapaPipelineAttr(S.Context, AL, ii));
+}
+
+static void handleTapaUnrollAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  S.Diag(D->getBeginLoc(), diag::warn_attribute_type_not_supported)
+      << AL << "unsupported unroll target";
+}
+
+static void handleTapaTargetAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  StringRef TargetStr;
+  SourceLocation ArgLoc;
+
+  if (AL.getNumArgs() == 0)
+    TargetStr = "xilinx_hls";
+  else if (!S.checkStringLiteralArgumentAttr(AL, 0, TargetStr, &ArgLoc))
+    return;
+
+  TapaTargetAttr::TargetType Target;
+  if (!TapaTargetAttr::ConvertStrToTargetType(TargetStr, Target)) {
+    S.Diag(AL.getLoc(), diag::warn_attribute_type_not_supported)
+        << AL << TargetStr << ArgLoc;
+    return;
+  }
+
+  D->addAttr(::new (S.Context) TapaTargetAttr(S.Context, AL, Target));
+}
+
+// End TAPA
+
 /// Check if the passed-in expression is of type int or bool.
 static bool isIntOrBool(Expr *Exp) {
   QualType QT = Exp->getType();
@@ -8033,6 +8070,19 @@
   case ParsedAttr::AT_GCCStruct:
     handleGCCStructAttr(S, D, AL);
     break;
+
+  // Tapa attributes.
+  case ParsedAttr::AT_TapaPipeline:
+    handleTapaPipelineAttr(S, D, AL);
+    break;
+
+  case ParsedAttr::AT_TapaUnroll:
+    handleTapaUnrollAttr(S, D, AL);
+    break;
+
+  case ParsedAttr::AT_TapaTarget:
+    handleTapaTargetAttr(S, D, AL);
+    break;
   }
 }
 
diff --git a/clang/lib/Sema/SemaStmtAttr.cpp b/clang/lib/Sema/SemaStmtAttr.cpp
--- a/clang/lib/Sema/SemaStmtAttr.cpp
+++ b/clang/lib/Sema/SemaStmtAttr.cpp
@@ -188,6 +188,50 @@
   return LoopHintAttr::CreateImplicit(S.Context, Option, State, ValueExpr, A);
 }
 
+// Begin TAPA
+
+static Attr *handleTapaPipelineAttr(Sema &S, Stmt *St, const ParsedAttr &A,
+                                    SourceRange) {
+  uint32_t ii = 0;
+  if (A.getNumArgs()) {
+    if (!S.checkUInt32Argument(A, A.getArgAsExpr(0), ii))
+      return nullptr;
+  }
+
+  if (St->getStmtClass() != Stmt::DoStmtClass &&
+      St->getStmtClass() != Stmt::ForStmtClass &&
+      St->getStmtClass() != Stmt::CXXForRangeStmtClass &&
+      St->getStmtClass() != Stmt::WhileStmtClass) {
+    S.Diag(St->getBeginLoc(), diag::warn_attribute_type_not_supported)
+        << A << "unsupported pipeline target";
+    return nullptr;
+  }
+
+  return ::new (S.Context) TapaPipelineAttr(S.Context, A, ii);
+}
+
+static Attr *handleTapaUnrollAttr(Sema &S, Stmt *St, const ParsedAttr &A,
+                                  SourceRange) {
+  uint32_t factor = 0;
+  if (A.getNumArgs()) {
+    if (!S.checkUInt32Argument(A, A.getArgAsExpr(0), factor))
+      return nullptr;
+  }
+
+  if (St->getStmtClass() != Stmt::DoStmtClass &&
+      St->getStmtClass() != Stmt::ForStmtClass &&
+      St->getStmtClass() != Stmt::CXXForRangeStmtClass &&
+      St->getStmtClass() != Stmt::WhileStmtClass) {
+    S.Diag(St->getBeginLoc(), diag::warn_attribute_type_not_supported)
+        << A << "unsupported unroll target";
+    return nullptr;
+  }
+
+  return ::new (S.Context) TapaUnrollAttr(S.Context, A, factor);
+}
+
+// End TAPA
+
 namespace {
 class CallExprFinder : public ConstEvaluatedExprVisitor<CallExprFinder> {
   bool FoundAsmStmt = false;
@@ -724,6 +768,10 @@
     return S.CreateAnnotationAttr(A);
   case ParsedAttr::AT_Atomic:
     return handleAtomicAttr(S, St, A, Range);
+  case ParsedAttr::AT_TapaPipeline:
+    return handleTapaPipelineAttr(S, St, A, Range);
+  case ParsedAttr::AT_TapaUnroll:
+    return handleTapaUnrollAttr(S, St, A, Range);
   default:
     if (Attr *AT = nullptr; A.getInfo().handleStmtAttribute(S, St, A, AT) !=
                             ParsedAttrInfo::NotHandled) {
